homework_solution.md

```diff
@@ -0,0 +1,474 @@
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id;
+      ```
+
+   4. Insert into the `stations` table the result of our query
+
+      ```
+      INSERT INTO stations (
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+         UNION
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+      );
+      ```
+
+   5. Adjust the `id` sequence to be the highest `id` from our `INSERT` above
+
+      ```
+      SELECT setval(pg_get_serial_sequence('stations', 'id'), (SELECT MAX(id) FROM stations));
+      ```
+
+   6. Finally, add entries for records where a name exists, but the id is null, relying on the seqquence to generate an id. This has to be done in 2 steps, so we can exclude the records from the first insertions when we run the second one, avoiding duplicates.
+
+      ```
+      INSERT INTO stations (
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), from_station_name
+         FROM trips
+         WHERE
+            from_station_id IS NULL
+         AND
+            from_station_name NOT IN (SELECT name from stations)
+         GROUP BY from_station_name
+      );
+      INSERT INTO stations(
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), to_station_name
+         FROM trips
+         WHERE
+            to_station_id IS NULL
+         AND
+            to_station_name NOT IN (SELECT name from stations)
+         GROUP BY to_station_name
+      );
+      ```
+
+2. _**Should we add any indexes to the stations table, why or why not?**_
+
+   Not right now. We already have an index on `id`, because it's the table's pkey. For our current usage, that should suffice.
+
+3. _**Fill in the missing data in the `trips` table based on the work you did above**_
+
+   ### Steps
+
+   1. We _could_ update the names in `trips`, like so:
+
+      ```
+      UPDATE trips t
+      SET
+         from_station_name = (SELECT name FROM stations s WHERE t.from_station_id = s.id),
+         to_station_name = (SELECT name FROM stations s WHERE t.from_station_id = s.id)
+      ```
+
+   2. However, it's best to just delete the station names column, so the `stations` table becomes the single source of truth for that (but we won't do that just yet!!)
+
+      ```
+      ALTER TABLE trips_backup
+      DROP COLUMN from_station_name,
+      DROP COLUMN to_station_name;
+      ```
+
+   3. Before dropping the columns, we have to populate the station ids columns where they're null
+
+      ```
+      UPDATE trips t
+      SET from_station_id = (SELECT id FROM stations s WHERE t.from_station_name = s.name),
+      to_station_id  = (SELECT id FROM stations s WHERE t.to_station_name = s.name)
+      WHERE from_station_id IS NULL OR to_station_id IS NULL
+      ```
```

Comment:
[question]
What will happen if a station name is encountered here that doesn't exist in the stations table? How could this happen?

Reply:
It could happen if we didn't execute the insertion into `stations` of records where just the name was present.
In this case, that query had an error, and we skipped the `to_station` portion of it, but we were lucky to have enough overlap with the `from_station` that it covered all of our needs.